compose/snippets/src/main/java/com/example/compose/snippets/pictureInPicture/PictureInPictureSnippets.kt

```diff
@@ -0,0 +1,384 @@
+        }
+    }
+    // [END android_compose_pip_post_12_should_enter_pip]
+}
+
+@Composable
+fun PipListenerPreAPI12_1(shouldEnterPipMode: Boolean) {
+    // [START android_compose_pip_pre12_should_enter_pip]
+    val currentShouldEnterPipMode by rememberUpdatedState(newValue = shouldEnterPipMode)
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O &&
+        Build.VERSION.SDK_INT < Build.VERSION_CODES.S
+    ) {
+        val context = LocalContext.current
+        DisposableEffect(context) {
+            val onUserLeaveBehavior: () -> Unit = {
+                if (currentShouldEnterPipMode) {
+                    context.findActivity()
+                        .enterPictureInPictureMode(PictureInPictureParams.Builder().build())
+                }
+            }
+
+            // [END android_compose_pip_pre12_should_enter_pip]
+            onDispose {
+            }
+        }
+    }
+}
+
+@Composable
+fun VideoPlayer1(
+    shouldEnterPipMode: Boolean,
+    modifier: Modifier = Modifier,
+) {
+    // [START android_compose_pip_set_source_rect]
+    val context = LocalContext.current
+
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+        val pipModifier = modifier.onGloballyPositioned { layoutCoordinates ->
+            val builder = PictureInPictureParams.Builder()
+            if (shouldEnterPipMode) {
+                val sourceRect = layoutCoordinates.boundsInWindow().toAndroidRectF().toRect()
+                builder.setSourceRectHint(sourceRect)
+            }
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+                builder.setAutoEnterEnabled(shouldEnterPipMode)
+            }
+            context.findActivity().setPictureInPictureParams(builder.build())
+        }
+    }
+    // [END android_compose_pip_set_source_rect]
+}
+
+@Composable
+fun VideoPlayer2(
+    shouldEnterPipMode: Boolean,
+    modifier: Modifier = Modifier,
+) {
+    // [START android_compose_pip_set_aspect_ratio]
+    val context = LocalContext.current
+
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+        val pipModifier = modifier.onGloballyPositioned { layoutCoordinates ->
+            val builder = PictureInPictureParams.Builder()
+
+            if (shouldEnterPipMode) {
+                val sourceRect = layoutCoordinates.boundsInWindow().toAndroidRectF().toRect()
+                builder.setSourceRectHint(sourceRect)
+                builder.setAspectRatio(
+                    Rational(sourceRect.width(), sourceRect.height())
+                )
+            }
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+                builder.setAutoEnterEnabled(shouldEnterPipMode)
+            }
+            context.findActivity().setPictureInPictureParams(builder.build())
+        }
+    }
+    // [END android_compose_pip_set_aspect_ratio]
+}
+
+// [START android_compose_pip_build_remote_action]
+@RequiresApi(Build.VERSION_CODES.O)
+private fun buildRemoteAction(
```

Comment:
Not sure I see the value of this function - why not create a `RemoteAction` directly?